File: examples/Example_37.java
/*
 * Example_37.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package examples;

import java.io.*;
import java.util.*;
import com.pdfjet.*;
import com.pdfjet.fonts.*;

/**
 * Example_37.java
 */
class Example_37 {
    public Example_37(String fileName) throws Exception {
        PDF pdf = new PDF(new BufferedOutputStream(new FileOutputStream("Example_37.pdf")));
        List<PDFobj> objects = pdf.read(new FileInputStream(fileName));

        Font f1 = new Font(
                objects,
                new FileInputStream(IBMPlexSans.Regular));
        f1.setSize(72f);

        TextLine text = new TextLine(f1, "This is a test!");
        text.setLocation(150f, 350f);
        text.setTextColor(Color.peru);

        List<PDFobj> pages = pdf.getPageObjects(objects);
        for (PDFobj pageObj : pages) {
            GraphicsState gs = new GraphicsState();
            gs.setAlphaStroking(0.75f);         // Stroking alpha
            gs.setAlphaNonStroking(0.75f);      // Non-stroking alpha
            pageObj.setGraphicsState(gs, objects);

            Page page = new Page(pdf, pageObj);
            page.addResource(f1, objects);
            page.setBrushColor(Color.blue);
            // page.drawString(f1, "Hello, World!", 50f, 200f);
            text.drawOn(page);

            page.complete(objects); // The graphics stack is unwinded automatically
        }
        pdf.addObjects(objects);

        pdf.complete();
    }

    public static void main(String[] args) throws Exception {
        long time0 = System.currentTimeMillis();
        new Example_37("data/testPDFs/wirth.pdf");
        long time1 = System.currentTimeMillis();
        System.out.printf("Example_37 => %4d ms%n", time1 - time0);
    }
}   // End of Example_37.java

File: examples/Example_37.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Diagnostics;
using PDFjet.NET;

/**
 *  Example_37.cs
 */
class Example_37 {
    public Example_37() {
        PDF pdf = new PDF(new BufferedStream(
                new FileStream("Example_37.pdf", FileMode.Create)));

        FileStream fis = new FileStream("data/testPDFs/wirth.pdf", FileMode.Open, FileAccess.Read);
        // FileStream fis = new FileStream("../../eBooks/UniversityPhysicsVolume1.pdf", FileMode.Open, FileAccess.Read);
        // FileStream fis = new FileStream("data/testPDFs/Smalltalk-and-OO.pdf", FileMode.Open, FileAccess.Read);
        // FileStream fis = new FileStream("data/testPDFs/InsideSmalltalk1.pdf", FileMode.Open, FileAccess.Read);
        // FileStream fis = new FileStream("data/testPDFs/InsideSmalltalk2.pdf", FileMode.Open, FileAccess.Read);
        // FileStream fis = new FileStream("data/testPDFs/Greenbook.pdf", FileMode.Open);
        // FileStream fis = new FileStream("data/testPDFs/Bluebook.pdf", FileMode.Open);
        // FileStream fis = new FileStream("data/testPDFs/Orangebook.pdf", FileMode.Open);

        List<PDFobj> objects = pdf.Read(fis);
        fis.Close();

        Font f1 = new Font(objects,
                new FileStream("fonts/OpenSans/OpenSans-Regular.ttf.stream",
                FileMode.Open,
                FileAccess.Read), Font.STREAM);
        f1.SetSize(72f);

        TextLine line = new TextLine(f1, "This is a test!");
        line.SetLocation(50f, 350f);
        line.SetColor(Color.peru);

        List<PDFobj> pages = pdf.GetPageObjects(objects);
        foreach (PDFobj pageObj in pages) {
            GraphicsState gs = new GraphicsState();
            gs.SetAlphaStroking(0.75f);         // Set alpha for stroking operations
            gs.SetAlphaNonStroking(0.75f);      // Set alpha for nonstroking operations
            pageObj.SetGraphicsState(gs, objects);

            Page page = new Page(pdf, pageObj);
            page.AddResource(f1, objects);
            page.SetBrushColor(Color.blue);
            page.DrawString(f1, "Hello, World!", 50f, 200f);

            line.DrawOn(page);

            page.Complete(objects); // The graphics stack is unwinded automatically
        }
        pdf.AddObjects(objects);
/*
        List<Image> images = new List<Image>();
        foreach (PDFobj obj in objects.Values) {
            if (obj.GetValue("/Subtype").Equals("/Image")) {
                float w = float.Parse(obj.GetValue("/Width"));
                float h = float.Parse(obj.GetValue("/Height"));
                if (w > 500f && h > 500f) {
                    images.Add(new Image(pdf, obj));
                }
            }
        }

        Font f1 = new Font(pdf, CoreFont.HELVETICA);
        f1.SetSize(72f);

        Page page = null;
        foreach (Image image in images) {
            page = new Page(pdf, A4.PORTRAIT);

            GraphicsState gs = new GraphicsState();
            gs.Set_CA(0.7f);    // Stroking alpha
            gs.Set_ca(0.7f);    // Nonstroking alpha
            page.SetGraphicsState(gs);

            image.ResizeToFit(page, true);

            // image.FlipUpsideDown(true);
            // image.SetLocation(0f, -image.GetHeight());

            // image.RotateClockwise(180);
            // image.SetLocation(0f, 0f);

            image.DrawOn(page);

            TextLine text = new TextLine(f1, "Hello, World!");
            text.SetColor(Color.blue);
            text.SetLocation(150f, 150f);
            text.DrawOn(page);

            page.SetGraphicsState(new GraphicsState());
        }
*/
        pdf.Complete();
    }

    public static void Main(String[] args) {
        Stopwatch sw = Stopwatch.StartNew();
        long time0 = sw.ElapsedMilliseconds;
        new Example_37();
        long time1 = sw.ElapsedMilliseconds;
        TextUtils.PrintDuration("Example_37", time0, time1);
    }
}   // End of Example_37.cs
